### DISCUSSION TITLE
API: SvelteFlowProvider, nodes and edges

### DISCUSSION THREAD

**Discussion by moklick** [xyflow/xyflow#2971] [Svelte]
Hey! 

we are currently working on a port to Svelte: SvelteFlow.

There is one API decision I am still not sure about and that I would like to improve. In SvelteFlow you create custom stores for nodes and edges in user land and pass them to the `SvelteFlowProvider`:

**Flow.svelte**
```svelte 
<script lang="ts">
  const nodes = createNodes([
    {
      id: '1',
      type: 'input',
      data: { label: 'Input Node' },
      position: { x: 150, y: 5 }
    },
    {
      id: '2',
      type: 'default',
      data: { label: 'Node' },
      position: { x: 0, y: 150 }
    }
  ]);

  const edges = createEdges([
    {
      id: '1-2',
      type: 'default',
      source: '1',
      target: '2',
      label: 'Edge Text'
    }
  ]);
</script>

<SvelteFlowProvider
  {nodes}
  {edges}
>
  <!-- in this component we want to access the viewport, nodes and edges --> 
  <ViewportDisplay />
  <SvelteFlow
    fitView
    minZoom={0.1}
    maxZoom={2.5}
  >
    <Controls />
    <Background />
    <Minimap />
  </SvelteFlow>
</SvelteFlowProvider>
```

**ViewportDisplay.svelte** 
```svelte 
<script lang="ts">
  const { zoomIn, setCenter } = useSvelteFlow();
</script>

<button on:click={() => zoomIn()}>zoom in</button>
<button on:click={() => setCenter(0, 0)}>setCenter 0, 0</button>
```

The custom nodes and edges stores get directly used by the "global store". The global store is created by the [SvelteFlowProvider](https://github.com/wbkd/react-flow/blob/feat/packages/packages/svelte/src/lib/components/SvelteFlowProvider/SvelteFlowProvider.svelte#L12-L19) and added to a context, so that it can be used in every component inside the provider. This is OK, but I would really try to find a way how I could pass the nodes and edges to the `SvelteFlow` component and still be able to use them in components that are higher in the component tree. If anyone can help out here or have an idea, please let me know :)

Some context:

- we use custom stores for nodes and edges to prevent the need to sync the user state with the one of SvelteFlow. SvelteFlow updates the position and other attributes. With this solution it's easy to subscribe or update nodes and edges in user land.
- we have a hook called `useSvelteFlow` that exports helpers like `zoomTo`, `setCenter` and so on. It must be possible to use this hook in components that are higher in the component tree than the `SvelteFlow` component. That's why we need the provider. In the example above it could be the `<ViewportDisplay />` component where we would like to use that hook. 
- in the example above we can pass `minZoom` and `maxZoom` to the `SvelteFlow` component and still access those values in the `<ViewportDisplay />` component. This is possible, because `minZoom` and `maxZoom` are primitives that are used to overwrite the store values of the minZoomStore and maxZoomStore. Unfortunately this is not possible with the custom nodes and edges stores because we need to use the actual stores created by the user.

**Comment by braebo** [comment]
Could you use `context="module"` and the `export` keyword for this?

`Example.svelte`
```html
<script lang="ts" context="module">
  export const nodes = createNodes([/*...*/]);

  export const edges = createEdges([/*...*/]);
</script>

<SvelteFlowProvider
  {nodes}
  {edges}
>
  <!-- in this component we want to access the viewport, nodes and edges --> 
  <ViewportDisplay />
  <!-- etc.. --> 
</SvelteFlowProvider>
```

`Elsewhere.svelte`

```html
<script lang="ts">
  import { nodes, edges } from '$lib/Example.svelte'
  
  console.log(nodes, edges)
</script>
```

**Reply by moklick** [reply]
This would work for nodes and edges (custom stores defined by the user) but not for the other things like the viewport, or helper functions that are part of the context.

**Comment by PapaTemporal** [comment]
I might be mis-interpreting the question here, but I think what you are looking for is slot props. https://svelte.dev/tutorial/slot-props

Slot props allow you to assign a variable from a child to the container so it can be used in the parent.

**Reply by PapaTemporal** [reply]
Also, have you seen this: https://github.com/sidharth-anand/svelte-flow? It's no longer maintained, but very similar to react-flow and has almost the same inner workings already.

**Reply by moklick** [reply]
Yep, I saw that project. It seems like a 1:1 copy of React Flow maybe for learning purposes?! 

Thanks for mentioning the slot props but I am not sure how that could help here.

**Reply by PapaTemporal** [reply]
It could be for educational purposes. Was just curious as to if it could help to see how they did it.

As far as the props piece, I totally mis-interpreted the question. I was thinking that the nodes and edges would be created in SvelteFlow and exposed to the parent using the binding. But the scope won't be right, sorry.

**Reply by moklick** [reply]
Thanks, got it. Yeah it's also a bit hard for me to explain it well it seems :D

**Comment by moklick** [comment]
I ended up syncing nodes store that is passed by the user and the internal one. The API will look like this:

```svelte
<script lang="ts">
  const nodes = writable([
    {
      id: '1',
      type: 'input',
      data: { label: 'Input Node' },
      position: { x: 150, y: 5 }
    },
    {
      id: '2',
      type: 'default',
      data: { label: 'Node' },
      position: { x: 0, y: 150 }
    }
  ]);

  const edges = writable([
    {
      id: '1-2',
      type: 'default',
      source: '1',
      target: '2',
      label: 'Edge Text'
    }
  ]);
</script>

<SvelteFlowProvider>
  <ViewportDisplay />
  <SvelteFlow
    {nodes}
    {edges}
    fitView
    minZoom={0.1}
    maxZoom={2.5}
  >
    <Controls />
    <Background />
    <Minimap />
  </SvelteFlow>
</SvelteFlowProvider>
```

**Comment by KaiyuanMa** [comment]
This is sick! Any updates?

**Reply by PapaTemporal** [reply]
Take a look at the xyflow branch. Has the latest on the svelteflow setup. But I haven't seen an official release yet.
